flake-modules/github-releases.nix

```diff
@@ -0,0 +1,132 @@
+            By default the GitHub pages deployment is triggered by the `onPush.default` job,
+            so that the deployment only proceeds when the default builds are successful.
+          '';
+          default = "default";
+          defaultText = lib.literalExpression "default";
+        };
+      };
+    };
+
+  config =
+    let
+      inherit (lib) mkIf mkMerge;
+      inherit (config) defaultEffectSystem;
+
+      cfg = config.hercules-ci.github-releases;
+      enable = cfg.files != [];
+    in
+    mkIf enable {
+      herculesCI = herculesCI@{ config, ... }:
+        let
+          deploy = withSystem defaultEffectSystem ({ hci-effects, ... }:
+            hci-effects.modularEffect {
+              imports = [
+                ../effects/modules/git-auth-gh.nix
+              ];
+              secretsMap = {
+                token = { type = "GitToken"; };
+              };
+              git.checkout = {
+                remote.url = config.repo.remoteHttpUrl;
+                forgeType = config.repo.forgeType;
+              };
+              effectScript = ''
+                gh repo clone ${config.repo.owner}/${config.repo.name} source -- --branch ${config.repo.tag} --single-branch
+                cd source
+                gh release create \
+                  --verify-tag ${cfg.releaseTag herculesCI} \
+                  ${lib.concatStringsSep " " cfg.files}
+              '';
+              env = { };
+            }
+          );
+        in
+        {
+          onPush = mkMerge [
+            {
+              ${cfg.pushJobName}.outputs.effects.gh-releases =
+                lib.optionalAttrs
+                  (cfg.condition herculesCI.config.repo)
+                  deploy;
+            }
+            {
+              default.outputs.checks.release-artifacts = mkIf cfg.checkArtifacts (withSystem defaultEffectSystem ({ pkgs, ... }:
+                let artifacts-checker = pkgs.writers.writePython3 "artifacts-checker" {} ''
+                      from os import environ
+                      from os.path import isfile, realpath
+
+                      for file in environ["FILES"].split("\n"):
+                          path, *labelMaybe = file.rsplit("#", maxsplit=1)
+                          labelMessage = labelMaybe and f"with label `{labelMaybe[0]}` " or ""
+                          print(f"Checking that path {path} {labelMessage}exists: ", end="")
+                          try:
+                              rpath = realpath(path, strict=True)
+                              if not isfile(rpath):
+                                  print("Not a file")
+                                  exit(1)
+                          except Exception as e:
+                              print(f"Cannot access {path}")
+                              raise e
+                          print("OK")
+
+                      open(environ["out"], "w")
+                    '';
```

Comment:
Hmm, I was kind of expecting this to be easier. Now I'm thinking if we're getting python involved, we might as well stay in JSON land; take JSON as input and construct call `gh` from the script, unless in check mode.

JSON input would let us declare releases with submodules

```nix
[ {
  filename = "dist.tar.gz"; #optional
  label = "Dist Tarball"; # optional
  contents = distTarball;
} ]
```

It appears that filenames have to be unique, but an attribute set would make one of the optional fields required, so let's stick with a list.

If that's not something you wish to implement, I think this could be made more accurate by passing the input via arguments instead of an environment variable. That will slightly simplify the splitting and restore some support for globbing.

Reply:
Oh, good idea. Label *is* the filename though, it's not an additional information. I propose using `Array<Path | {path: Path, label: string}>` (in Typescript notation). Second variant translates to `path#label`, first one to just `path`.

Reply:
This format potentially allows the check to be split along the derivation `system`, as an optimization and to run in `nix flake check` via `perSystem`. Grouping and the potential for a remainder aren't trivial, so this might be a bit excessive (but nice).

Reply:
> `Array<Path | {path: Path, label: string}>`

Lgtm!
The triple I suggested probably requires a file rename anyway. It's ok to push that rename to the user.